### ISSUE TITLE
Context with timeout not cancelling database call

### ISSUE THREAD

**Issue by veqryn** [go-goracle/goracle#48]
We are calling a stored procedure like so:
```go
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
	defer cancel()

	_, err := db.updateLastScheduledStmt.ExecContext(ctx, mb.ID)
	return err
```

This about 1-10 times per day, this call never returns for unknown reasons, causing our process to just hang there until our cluster kills the whole pod 30 minutes later for failing a health check.

Not sure if it matters, but `updateLastScheduledStmt` is a `*sqlx.Stmt`, which was successfully prepared and used successfully in previous calls.  This is in the hot path of this app, and this call is being made about once per second or more, 100% successfully except when it doesn't return those few times per day.

I don't know if the issue is specific to prepared stored procedures, or if timeout-contexts generally fail to cancel all database calls.

I am using git commit 94abed0e1b9d25c4e97186353f37a8825f02a704 of goracle.

**Comment by tgulacsi** [comment]
I see no obvious reason - I've moved another C calling code behind the timeout barrier, this may help.
Please try 174f046 and configure logging for that call!

**Comment by veqryn** [comment]
FYI: strings.Builder is a golang 1.10+ construct, and since I am on golang 1.9 til the first point release, i get this:
`go/src/github.com/ReturnPath/seed-backend/vendor/gopkg.in/goracle.v2/rows.go:302: undefined: strings.Builder`

**Comment by veqryn** [comment]
So, after updating goracle and adding a logger, it still happens.  Here are my logs:
```
{"t":"2018-03-02T17:27:10.146Z","l":"debug","m":"mark_retriever_scheduled","f":{"mailbox":1636300022}}
{"t":"2018-03-02T17:27:10.146Z","l":"debug","m":"db","f":{"v":"[enter bindVars args [{ 1 1636300022}]]"}}
{"t":"2018-03-02T17:27:10.146Z","l":"debug","m":"db","f":{"v":"[msg bindVars i 0 in true out false value int64 1636300022]"}}
{"t":"2018-03-02T17:27:10.146Z","l":"debug","m":"db","f":{"v":"[doManyCount 1 arrLen -1 doExecMany false minArrLen maxArrLen]"}}
{"t":"2018-03-02T17:27:10.146Z","l":"debug","m":"db","f":{"v":"[msg newVar i 0 plSQLArrays false typ 2010 natTyp 3000 sliceLen 1 bufSize 0 isSlice false]"}}
{"t":"2018-03-02T17:27:10.146Z","l":"debug","m":"db","f":{"v":"[msg set i 0 value int64=1636300022]"}}
{"t":"2018-03-02T17:27:10.146Z","l":"debug","m":"db","f":{"v":"[C dpiStmt_execute mode 32 colCount 0]"}}
<20 minutes go by, then we kill the container>
```

For reference, here is what it looks like when it is successful (basically the same):
```
{"t":"2018-03-02T17:27:10.066Z","l":"debug","m":"mark_retriever_scheduled","f":{"mailbox":1956700022}}
{"t":"2018-03-02T17:27:10.066Z","l":"debug","m":"db","f":{"v":"[enter bindVars args [{ 1 1956700022}]]"}}
{"t":"2018-03-02T17:27:10.066Z","l":"debug","m":"db","f":{"v":"[msg bindVars i 0 in true out false value int64 1956700022]"}}
{"t":"2018-03-02T17:27:10.066Z","l":"debug","m":"db","f":{"v":"[doManyCount 1 arrLen -1 doExecMany false minArrLen maxArrLen]"}}
{"t":"2018-03-02T17:27:10.066Z","l":"debug","m":"db","f":{"v":"[msg newVar i 0 plSQLArrays false typ 2010 natTyp 3000 sliceLen 1 bufSize 0 isSlice false]"}}
{"t":"2018-03-02T17:27:10.066Z","l":"debug","m":"db","f":{"v":"[msg set i 0 value int64=1956700022]"}}
{"t":"2018-03-02T17:27:10.066Z","l":"debug","m":"db","f":{"v":"[C dpiStmt_execute mode 32 colCount 0]"}}
{"t":"2018-03-02T17:27:10.137Z","l":"debug","m":"Sending mbox to SQS","f":{"mbox":{"id":1956700022}}}
```

**Comment by tgulacsi** [comment]
Regarding `mark_retriever_scheduled`: who commits?

Back to the hangup: The "C", "dpiStmt_execute" is printed right before calling C.dpiStmt_execute.
Right after it it would log the error (ok, move the "if err==nil {break}" part AFTER the logging).

So it seems it hangs on C side.

The strange thing is that it seems it does not call Break (and print "BREAK") in the other goroutine.

[UPDATE 1]: I've created a test for this break functionality, and it fails...

**Comment by veqryn** [comment]
Thank you @tgulacsi , I will try this out and get back to you on the results.

I do have a few questions about the mutex locking in this driver:

First: It looks like with the change to a RW Mutex, there is a slight possibility of a deadlock:
1. User calls ExecContext, which acquires a Read Lock and then waits on the db
2. User then calls PrepareContext or Commit or something in another goroutine, which attempts to acquire a Write Lock, and blocks
3. First call to ExecContext hits the context timeout, calling Break, which attempts to acquire a Read Lock, but is now blocked by the second call to PrepareContext (RWMutex blocked Lock calls will block RLock calls)

Second: It looks like some of the calls for Write Lock are for the setting of trace tags:
```go
	c.Lock()
	if tt, ok := ctx.Value(traceTagCtxKey).(TraceTag); ok {
		c.setTraceTag(tt)
	}
	c.Unlock()
```
I could be wrong, but I think you could lock only if the tag exists:
```go
	if tt, ok := ctx.Value(traceTagCtxKey).(TraceTag); ok {
		c.Lock()
		c.setTraceTag(tt)
		c.Unlock()
	}
```

Third: Why to statements acquire and hold locks in order to exec?
This would mean multiple calls to execute a prepared statement would all contend with each other.
I would have thought this would be something that the database takes care of, and we wouldn't have to lock or need a mutex in the driver for this.
```gofunc (st *statement) ExecContext(ctx context.Context, args []driver.NamedValue) (driver.Result, error) {
	st.Lock()
	defer st.Unlock()
	st.conn.RLock()
	defer st.conn.RUnlock()
```

**Comment by tgulacsi** [comment]
You're partially right: I'll change setTraceTag locking.
But it's been using Mutex, not this more granular RWMutex.

I've tried to use Lock where *conn receives a modification; RLock where it is only read.

AFAIK database/sql will use a different Conn (thus Oracle session) for the PrepareContext after ExecContext.

But locking is hard, so I'll accept tests for such edge cases....

**Comment by tgulacsi** [comment]
You're partially right: I'll change setTraceTag locking.
But it's been using Mutex, not this more granular RWMutex.
Since 22d51006a5d6d89d3870843c3cf868d285c58da3 Break, Ping, BeginTx, ServerVersion, ExecContext, QueryContext became RLock from Lock beforehand. So locking became less strict.

I've tried to use Lock where *conn receives a modification; RLock where it is only read.

AFAIK database/sql will use a different Conn (thus Oracle session) for the PrepareContext after ExecContext.

But locking is hard, so I'll accept tests for such edge cases....

**Comment by veqryn** [comment]
hm... I updated to 6a7f64da825a70d48f3dc46a3bc054dd7fb8edfd this morning, and when I checked just now we had 8 pod restarts, with the context 5 minute timeout still not working.

```
{"t":"2018-03-04T04:37:09.028Z","l":"debug","m":"mark_retriever_scheduled","f":{"mailbox":1977800022}}
{"t":"2018-03-04T04:37:09.028Z","l":"debug","m":"db","f":{"v":"[enter bindVars args [{ 1 1977800022}]]"}}
{"t":"2018-03-04T04:37:09.028Z","l":"debug","m":"db","f":{"v":"[msg bindVars i 0 in true out false value int64 1977800022]"}}
{"t":"2018-03-04T04:37:09.028Z","l":"debug","m":"db","f":{"v":"[doManyCount 1 arrLen -1 doExecMany false minArrLen maxArrLen]"}}
{"t":"2018-03-04T04:37:09.028Z","l":"debug","m":"db","f":{"v":"[msg newVar i 0 plSQLArrays false typ 2010 natTyp 3000 sliceLen 1 bufSize 0 isSlice false]"}}
{"t":"2018-03-04T04:37:09.028Z","l":"debug","m":"db","f":{"v":"[msg set i 0 value int64=1977800022]"}}
{"t":"2018-03-04T04:37:09.028Z","l":"debug","m":"db","f":{"v":"[C dpiStmt_execute mode 32 colCount 0]"}}
{"t":"2018-03-04T04:58:53.726Z","l":"info","m":"Caught cancellation signal; shutting down","f":{"signal":"terminated"}}
```

**Comment by tgulacsi** [comment]
Strange.
Does TestExecHang work in your env?
Can you copy the panic (sigHUP) stack trace here? That'd point exactly
where the program is waiting.
Or just sprinkling the code with Log calls before&after the break call?

Chris Duncan <notifications@github.com> ezt írta (időpont: 2018. márc. 4.,
V, 7:49):

**Comment by veqryn** [comment]
I'll try to add some logging and stuff tomorrow.  Maybe I can even catch it in the act and get a thread dump with gops.

**Comment by veqryn** [comment]
Ok, here is what I've got with tons of logging added:

```
{"t":"2018-03-06T00:44:40.627Z","l":"debug","m":"mark_retriever_scheduled","f":{"mailbox":1977200022}}
{"t":"2018-03-06T00:44:40.627Z","l":"debug","m":"db","f":{"v":"[msg ExecContext before lock]"}}
{"t":"2018-03-06T00:44:40.627Z","l":"debug","m":"db","f":{"v":"[msg ExecContext after lock]"}}
{"t":"2018-03-06T00:44:40.627Z","l":"debug","m":"db","f":{"v":"[enter bindVars args [{ 1 1977200022}]]"}}
{"t":"2018-03-06T00:44:40.627Z","l":"debug","m":"db","f":{"v":"[msg bindVars i 0 in true out false value int64 1977200022]"}}
{"t":"2018-03-06T00:44:40.627Z","l":"debug","m":"db","f":{"v":"[doManyCount 1 arrLen -1 doExecMany false minArrLen maxArrLen]"}}
{"t":"2018-03-06T00:44:40.627Z","l":"debug","m":"db","f":{"v":"[msg newVar i 0 plSQLArrays false typ 2010 natTyp 3000 sliceLen 1 bufSize 0 isSlice false]"}}
{"t":"2018-03-06T00:44:40.627Z","l":"debug","m":"db","f":{"v":"[msg set i 0 value int64=1977200022]"}}
{"t":"2018-03-06T00:44:40.628Z","l":"debug","m":"db","f":{"v":"[msg ExecContext entering first loop]"}}
{"t":"2018-03-06T00:44:40.628Z","l":"debug","m":"db","f":{"v":"[C dpiStmt_execute mode 32 colCount 0]"}}
{"t":"2018-03-06T00:49:40.627Z","l":"debug","m":"db","f":{"v":"[msg ExecContext context complete]"}}
{"t":"2018-03-06T00:49:40.628Z","l":"debug","m":"db","f":{"v":"[msg ExecContext context break calling]"}}
{"t":"2018-03-06T00:49:40.628Z","l":"debug","m":"db","f":{"v":"[msg Break before lock]"}}
{"t":"2018-03-06T00:49:40.628Z","l":"debug","m":"db","f":{"v":"[msg Break after lock]"}}
{"t":"2018-03-06T00:49:40.683Z","l":"debug","m":"db","f":{"v":"[msg Break before return]"}}
{"t":"2018-03-06T00:49:40.683Z","l":"debug","m":"db","f":{"v":"[msg ExecContext context break called]"}}
{"t":"2018-03-06T00:49:40.683Z","l":"debug","m":"db","f":{"v":"[msg ExecContext context break returned]"}}
```

Basically, it called `Break` successfully and that returned.
However, it looks like Break didn't actually do anything, and the original call `db.updateLastScheduledStmt.ExecContext(ctx, mb.ID)` still never returned.

Here are my threads:
```
goroutine 1 [syscall, 21 minutes, locked to thread]:
github.com/xxx/vendor/gopkg.in/goracle%2ev2._Cfunc_dpiStmt_execute(0x7f9c200cb040, 0x7f9c00000020, 0xc4203f7110, 0x0)
        github.com/xxx/vendor/gopkg.in/goracle.v2/_obj/_cgo_gotypes.go:2685 +0x4d
github.com/xxx/vendor/gopkg.in/goracle%2ev2.(*statement).ExecContext.func3(0x7f9c200cb040, 0x20, 0xc4203f7110, 0x0)
        /go/src/github.com/xxx/vendor/gopkg.in/goracle.v2/stmt.go:285 +0x72
github.com/xxx/vendor/gopkg.in/goracle%2ev2.(*statement).ExecContext(0xc42054e0f0, 0x10075c0, 0xc42049b320, 0xc4208ab890, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/xxx/vendor/gopkg.in/goracle.v2/stmt.go:285 +0x39c
database/sql.ctxDriverStmtExec(0x10075c0, 0xc42049b320, 0x1007900, 0xc42054e0f0, 0xc4208ab890, 0x1, 0x1, 0x1, 0x1, 0x0, ...)
        /usr/local/go/src/database/sql/ctxutil.go:66 +0x22e
database/sql.resultFromStatement(0x10075c0, 0xc42049b320, 0x1006640, 0xc4200e0d00, 0xc420282380, 0xc42019cc10, 0x1, 0x1, 0x0, 0x0, ...)
        /usr/local/go/src/database/sql/sql.go:2109 +0x184
database/sql.(*Stmt).ExecContext(0xc4201f70e0, 0x10075c0, 0xc42049b320, 0xc42019cc10, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/database/sql/sql.go:2085 +0x223
github.com/xxx/xxx/db.DB.UpdateLastScheduleDate(0xc420526c90, 0xc4203e6020, 0xc4203e6040, 0x75d9ad96, 0xc4208b0c80, 0x17, 0x0, 0xed22fdb29, 0x1036180, 0x0, ...)
        /go/src/github.com/xxx/db/mailbox.go:122 +0x1b5
<more of my packages>


goroutine 39 [select, 40 minutes, locked to thread]:
runtime.gopark(0xaaf2f8, 0x0, 0xa8a1b1, 0x6, 0x18, 0x1)
        /usr/local/go/src/runtime/proc.go:287 +0x12c
runtime.selectgo(0xc42004df50, 0xc4200ee3c0)
        /usr/local/go/src/runtime/select.go:395 +0x1149
runtime.ensureSigM.func1()
        /usr/local/go/src/runtime/signal_unix.go:511 +0x220
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:2337 +0x1


goroutine 42 [IO wait]:
internal/poll.runtime_pollWait(0x7f9c3b326eb0, 0x72, 0xffffffffffffffff)
        /usr/local/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc420248018, 0x72, 0xc420048c00, 0x0, 0x0)
        /usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/poll.(*pollDesc).waitRead(0xc420248018, 0xffffffffffffff00, 0x0, 0x0)
        /usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc420248000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/internal/poll/fd_unix.go:335 +0x1e2
net.(*ne/usr/local/go/src/net/fd_unix.go:238 +0x42048de8, 0x40b31b)
net.(*TC/usr/local/go/src/net/tcpsock_posix.go:136 +0x2e048e18, 0x40a0b7)
net.(*TC/usr/local/go/src/net/tcpsock.go:234 +0x4948e60, 0xc420048e68, 0xc420048e58)
net/http/usr/local/go/src/net/http/server.go:3120 +0x2feca8, 0xc42035c320, 0x1007600, 0xc420092750)
net/http/usr/local/go/src/net/http/server.go:2695 +0x1b209e058, 0x0, 0x0)
net/http/usr/local/go/src/net/http/server.go:2636 +0xa9a8ea0, 0x1002180)
net/http/usr/local/go/src/net/http/server.go:2882 +0x7fx0)
main.mai/go/src/github.com/xxx/seed_scheduler.go:43 +0x90
created /go/src/github.com/xxx/seed_scheduler.go:40 +0xd0
        /usr/local/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc420182318, 0x72, 0xffffffffffffff00, 0x1002bc0, 0xffda98)
internal/usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d00, 0x1000)
internal/usr/local/go/src/internal/poll/fd_unix.go:126 +0x18a0x1000, 0x0, 0x0, 0x0)
net.(*ne/usr/local/go/src/net/fd_unix.go:202 +0x5200, 0x1000, 0xc4205db8c8, 0x68c88a, 0xc4202a1178)
net.(*co/usr/local/go/src/net/net.go:176 +0x6dx1000, 0x1000, 0x0, 0x0, 0x0)
net/http/usr/local/go/src/net/http/server.go:753 +0x105 0x1000, 0x1000, 0xa87a59, 0xc4200882b0, 0xc4205dba20)
bufio.(*/usr/local/go/src/bufio/bufio.go:97 +0x11a
bufio.(*/usr/local/go/src/bufio/bufio.go:338 +0x2c0a, 0xc4205dba38, 0x0, 0xc4205dba38, 0xc4205dbab8, 0x41cf17)
bufio.(*/usr/local/go/src/bufio/bufio.go:367 +0x34, 0xa76120, 0x8, 0x22002022a700, 0xf8)
net/text/usr/local/go/src/net/textproto/reader.go:55 +0x705dbb20, 0xc4205dbb20, 0x41d678, 0x100, 0xa76120)
net/text/usr/local/go/src/net/textproto/reader.go:36 +0x2b0, 0x0, 0xc4205dbb90, 0x492ae2)
net/http/usr/local/go/src/net/http/request.go:925 +0x99x0, 0x0)
net/http/usr/local/go/src/net/http/server.go:933 +0x17cxc420105700, 0x0, 0x0, 0x0)
net/http/usr/local/go/src/net/http/server.go:1739 +0x50e5700)
created /usr/local/go/src/net/http/server.go:2720 +0x288
        /usr/local/go/src/database/sql/sql.go:871 +0x53
created by database/sql.Open40 minutes]:
database/usr/local/go/src/database/sql/sql.go:609 +0x1ee
        /usr/local/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc420248318, 0x72, 0xffffffffffffff00, 0x1002bc0, 0xffda98)
internal/usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d00, 0x1000)
internal/usr/local/go/src/internal/poll/fd_unix.go:126 +0x18a0x1000, 0x0, 0x0, 0x0)
net.(*ne/usr/local/go/src/net/fd_unix.go:202 +0x5200, 0x1000, 0xc4205d58c8, 0x68c88a, 0xc4202a0668)
net.(*co/usr/local/go/src/net/net.go:176 +0x6dx1000, 0x1000, 0x0, 0x0, 0x0)
net/http/usr/local/go/src/net/http/server.go:753 +0x105 0x1000, 0x1000, 0xa87a59, 0x3e002022a700, 0x398)
bufio.(*/usr/local/go/src/bufio/bufio.go:97 +0x11a
bufio.(*/usr/local/go/src/bufio/bufio.go:338 +0x2c0a, 0x41ca86, 0x7f9c3b3279f8, 0x4, 0xc4205d5ab8, 0x41cf17)
bufio.(*/usr/local/go/src/bufio/bufio.go:367 +0x34, 0xa76120, 0x1, 0x122002022a700, 0xf8)
net/text/usr/local/go/src/net/textproto/reader.go:55 +0x705d5b20, 0xc4205d5b20, 0x41d678, 0x100, 0xa76120)
net/text/usr/local/go/src/net/textproto/reader.go:36 +0x2b0, 0x0, 0xc4205d5b90, 0x492ae2)
net/http/usr/local/go/src/net/http/request.go:925 +0x99x0, 0x0)
net/http/usr/local/go/src/net/http/server.go:933 +0x17cxc4200a3700, 0x0, 0x0, 0x0)
net/http/usr/local/go/src/net/http/server.go:1739 +0x50e3700)
created /usr/local/go/src/net/http/server.go:2720 +0x288
```

**Comment by tgulacsi** [comment]
Thank you very much, this detail is invaluable!

Can you tweak TestExecHang in z_test.go to reproduce this strange phenomenon?
Can you ensure that st.Break is called on the same dpiConn as dpiStmt_execute is hung on? (%p)

But, from another view, what is your DB side process is doing? Maybe it's waiting on a lock?
(I want to reproduce it, that's why it would be good to know what kind of lock is it hung on).

I'll ask @anthony-tuininga of ODPI-C if he has any idea.

**Comment by veqryn** [comment]
1. TestExecHang passes for me. I suspect we'll have to keep digging to figure out why my stuff is failing to break properly, before we can create a proper test.

2. From looking at this driver's source, I don't see how anything else would be possible.
Is it possible that statement's `dpiStmt  *C.dpiStmt` is different than it statement's *conn's `dpiConn       *C.dpiConn`?

The application I am running is actually extremely simple.
It only opens one database connection, without any special parameters.
All action is taking place on a single thread/goroutine, with an infinite loop (no parallelism that i know of is happening, beyond whatever the driver and stdlib are doing under the hood).
We call a query to get a bunch of rows, and then we iterate through each of those items. For each we send a message to a queue and then call the above query that only sometimes hangs.

Is there some logging I should add to the driver to check for whether they use the same dpiConn?

3. Our database is under pretty heavy load recently, and there are a lot of queries (from other sources) that hit and query or update this table.  On top of this, we are increasingly split between multiple data-centers, and so there is some additional network latency that is probably affecting my application, as its oracle queries go across the country from the data center where my application is running to different data center where our Oracle DB is running.

I spoke with our DBA yesterday and they said two things to me:
First, they asked why my application had 3 sessions open (each of them was a "DIRECT" session, whatever that means).  I replied that I only had 1 instance of this application running, so it must be golang pooling connections or something.
Second, they said this query was causing some contention or something, but she wasn't sure why yet.

I'm honestly quite new to using Oracle, so I am not sure what questions to ask here.  I could use some specific directions:
What should I ask?
How else can I debug this on the client or server side?

**Comment by cemremengu** [comment]
Just curious @veqryn, are you creating a new sqlx.DB instance (open, connect etc.) everytime you call the SP? Is it possible that your DB is hitting max active sessions?

**Comment by veqryn** [comment]
No, I am reusing the same single connection made at the beginning. Besides, the prepared statements in go hold onto the connection that made them.

**Comment by cemremengu** [comment]
Do you think commiting inside the SP can make a difference? Sorry if these sound stupid, just trying to help by any means I can :)

**Comment by tgulacsi** [comment]
You haven't answered yet: who commits that UPDATE?

An open transaction can block other DML statements, and AFAIK even SELECTs!
Oracle Two-Phase commit can slow things down even more, and the default wait time for a hung transaction to commit is about 30min - but this is what your DBA should know better.

**Comment by veqryn** [comment]
@cemremengu 
That may help my database call, and I'll give it a try if everything else fails.
However, I actually am more interested in getting context timeouts to work properly, than I am in this particular db call.
Thank you for trying to help though.

@tgulacsi 
My understanding is that golang auto-commits all database calls that are not in a transaction.
I am not using transactions in this application, and so I am not explicitly calling commit anywhere in my code.  The updates being made do show up in the database, so I assume this is correct.

My DBA looked a bit further into this, and thinks the issue is most likely the wait events of: `SQL*Net message from client` (ie: network latency, broken tcp connection, etc) and/or Enq UL Contention (user lock).

Here is a screencap from one of the sessions:
![image](https://user-images.githubusercontent.com/1897829/37067366-a6db4a58-2166-11e8-83ab-4a61dbaf7b9c.png)

During one of the particular timeouts, we had several (unrelated) containers starting up and connecting to the database, giving a few TNS timeout errors:
```
Fatal NI connect error 12170.

 VERSION INFORMATION:
       TNS for Linux: Version 12.1.0.2.0 - Production
       Oracle Bequeath NT Protocol Adapter for Linux: Version 12.1.0.2.0 - Production
       TCP/IP NT Protocol Adapter for Linux: Version 12.1.0.2.0 - Production
 Time: 06-MAR-2018 15:58:27
 Tracing not turned on.
 Tns error struct:
   ns main err code: 12535

TNS-12535: TNS:operation timed out
   ns secondary err code: 12560
   nt main err code: 505

TNS-00505: Operation timed out
   nt secondary err code: 110
   nt OS err code: 0
 Client address: (ADDRESS=(PROTOCOL=tcp)(HOST=xx.xx.xx.xx)(PORT=xx))
Tue Mar 06 15:59:59 2018
Streams Capture C  3: first scn changed.
scn: 0x001f.8b41209a
Tue Mar 06 16:00:06 2018
```


Regardless of whether our packets are passing through several levels of Hell and going around the world twice on the way to oracle, or whatever else bad we might be doing, the driver should still respect the context timeout and successfully return from the call with an error.
I feel like it might be a problem with ODPI-C, since Break is being called.  But I also am just guessing.

Let me know if there is more debugging I can do to help.
I'll try to post a simplified version of our code also.

**Comment by tgulacsi** [comment]
I've filed https://github.com/oracle/odpi/issues/60 for advice.

**Comment by rbastic2** [comment]
Hi,

There are Oracle level parameters that may need to be configured. We have hit similar issues in production where the OCI-8 library does not properly execute a SQL*Net ORA break reset. I'm not clear on whether or not this is the same issue.

In production to solve this with a particularly troublesome network, we have the following parameters configured using a TNS file:

BREAK_POLL_SKIP=10000
SQLNET.SEND_TIMEOUT=10
SQLNET.RECV_TIMEOUT=10
SQLNET.INBOUND_CONNECT_TIMEOUT=10
SQLNET.OUTBOUND_CONNECT_TIMEOUT=10
SQLNET.EXPIRE_TIME=4
TCP.CONNECT_TIMEOUT=10
DISABLE_OOB=on

Please note that if there is any firewall rule or hardware filtering FIN/URG packets that "BREAK-RESETs" will end up filtered and this can trigger the type of freeze scenario being described.

If I am wrong then my apologies, it has been several months since we debugged this exact issue and it was a week of fighting Cisco routers, Oracle configuration, and so on, to sort this out. Please take this with a grain of salt.

I have another patch for reconnecting dropped connections that I will file in a separate issue. This may help in other ways, however, there are some trade-offs in problematic networks that I will touch on there.

**Comment by tgulacsi** [comment]
Thanks, @rbastic2 , hope this'll help @veqryn !

**Comment by tgulacsi** [comment]
@veqryn , you can (also) try execHang branch, it closes the conn after a context-deadline induced Break.

**Comment by veqryn** [comment]
@rbastic2 thanks for the advice.  I've spent more than an hour googling and can't figure out how or where to put those parameters.

Should they go into the CONNECT_DATA part of the connection description string?
```
myuser/mypassword@(DESCRIPTION =
	(ADDRESS_LIST =
		(ADDRESS = 
			(PROTOCOL = TCP)
			(HOST = myhost)
			(PORT = 1521)
		)
	)
	(CONNECT_DATA = 
		(SERVICE_NAME = myservicename)
		(SERVER = DEDICATED)
		(BREAK_POLL_SKIP = 10000)
		(SQLNET.SEND_TIMEOUT = 10)
		(SQLNET.RECV_TIMEOUT = 10)
		(SQLNET.INBOUND_CONNECT_TIMEOUT = 10)
		(SQLNET.OUTBOUND_CONNECT_TIMEOUT = 10)
		(SQLNET.EXPIRE_TIME = 4)
		(TCP.CONNECT_TIMEOUT = 10)
		(DISABLE_OOB = on)
	)
)
```

Can this be done in the connection string being passed to this driver?  I've much prefer not having to create files.
This is one of my least favorite parts of oracle.

**Comment by tgulacsi** [comment]
https://docs.oracle.com/cd/E11882_01/network.112/e10835/sqlnet.htm#NETRF182

These should go to the "sqlnet.ora" file.

Chris Duncan <notifications@github.com> ezt írta (időpont: 2018. márc. 7.,
Sze, 21:04):

**Comment by veqryn** [comment]
Alright, so I've put these settings into the `sqlnet.ora` file and deployed it.

I immediately started getting these errors:
```
{"t":"2018-03-07T21:52:00.604Z","l":"debug","m":"db.SelectMailBoxesToBeScheduled"}
{"t":"2018-03-07T21:52:00.875Z","l":"fatal","m":"Failed while running","e":"dpiStmt_execute: ORA-12609: TNS: Receive timeout occurred"}
```

As you can see, we issued a call, and less than 1 second later Oracle gives us this `Receive timeout` error.

I tried increasing the timeouts to more than 5 minutes (to be longer than my context timeout), but it doesn't matter.  The errors come back immediately, not after the number of seconds I have set.
```
DISABLE_OOB = on
BREAK_POLL_SKIP = 10
TCP.CONNECT_TIMEOUT = 390
SQLNET.RECV_TIMEOUT = 390
SQLNET.SEND_TIMEOUT = 390
SQLNET.INBOUND_CONNECT_TIMEOUT = 390
SQLNET.OUTBOUND_CONNECT_TIMEOUT = 390
SQLNET.EXPIRE_TIME = 8
```

**Comment by veqryn** [comment]
Well it seems after an initial bout of 10 or so crashes, the app stabilized and the above config (with 6.5 minute timeouts) is working.  I still do not know why it was giving `Receive timeout` before the configured timeout, but it doesn't seem to be doing it anymore for the last 3 hours.
Also to note, I added `(ENABLE=BROKEN)` to my connection string, which enables keep-alive packets.
After doing the sqlnet.ora file and the keep-alive packets, my context timeouts with breaks are now working again. 💯 

@rbastic2 I have a question about two of your sqlnet.ora options:
1. Why do you use `BREAK_POLL_SKIP = 10000`?  As far as I understand it, the only effect is the client spending less time looking for breaks/cancels/interrupts, which seems counter to what we would want?

2. Why do you use `DISABLE_OOB = on`?  This disables "Out of Band Breaks", which seem like a good thing, especially in my context above?


@tgulacsi I will try your new branch tomorrow. I had a quick look and I think switching which things happen in the new goroutine is a good idea overall.

**Comment by tgulacsi** [comment]
AFAIK DISABLE_OOB = on is against restrictive firewallsm

Chris Duncan <notifications@github.com> ezt írta (időpont: 2018. márc. 8.,
Cs 8:38):

**Comment by rbastic2** [comment]
I wish I could be of more help -- I don't have the answers though. Some of these were tuned by a dark mage, and I am but a lowly apprentice in Oracle's black arts.

Serious answer: this is a proprietary client environment with a troublesome network and certain no-nos being done. Firewall in-between the app and db, that sort of thing.

Even the FIN/URG issue I first mentioned wasn't well-understood by network engineers until packet sniffing made it clear what was happening between the Go application and the Oracle instance.

While I can't necessarily offer further advice for reasons of NDA, etc., on why we had to choose some of the settings (nor could I get that info at this point), I would advise you run some network analyzers while playing with these settings.

6.5 minutes is a pretty long time and you should be able to get that a lot lower if desired.

**Comment by rbastic2** [comment]
Wow. Ironically, I thought this problem was fixed on our end, but it has resurfaced. Any word on whether tgulacsi's branch is fixing it for you? I'll be trying to the ENABLE=BROKEN parameter tomorrow.

**Comment by veqryn** [comment]
I decided to compare the difference between running with and without the OOB settings for a day each first, before I try the branch.
I'll post more details tomorrow, but so far running without the OOB settings gives us totally different errors then the errors we get when running with the settings.

**Comment by rbastic2** [comment]
Interesting stuff and thanks for responding so quickly. Excited to read your next report!

**Comment by veqryn** [comment]
So here are my results from experimenting with `sqlnet.ora` file, combined with `(ENABLE=BROKEN)`
```
DISABLE_OOB = on
BREAK_POLL_SKIP = 10
TCP.CONNECT_TIMEOUT = 390
SQLNET.RECV_TIMEOUT = 390
SQLNET.SEND_TIMEOUT = 390
SQLNET.INBOUND_CONNECT_TIMEOUT = 390
SQLNET.OUTBOUND_CONNECT_TIMEOUT = 390
SQLNET.EXPIRE_TIME = 8
```
The timeouts are staggered so that I can see which ones have any effect.  My context timeout is 5 minutes, the send/receive timeouts are 6.5 minutes, and oracle confirms the connect is still active every 8 minutes.

With the above setup, my logs generally showed 2 cases coming up:
```
{"t":"2018-03-08T10:55:45.534Z","l":"debug","m":"db.UpdateLastScheduleDate"}
{"t":"2018-03-08T10:55:45.534Z","l":"debug","m":"db","f":{"v":"[enter bindVars args [{ 1 1386500022}]]"}}
{"t":"2018-03-08T10:55:45.534Z","l":"debug","m":"db","f":{"v":"[msg bindVars i 0 in true out false value int64 1386500022]"}}
{"t":"2018-03-08T10:55:45.534Z","l":"debug","m":"db","f":{"v":"[doManyCount 1 arrLen -1 doExecMany false minArrLen maxArrLen]"}}
{"t":"2018-03-08T10:55:45.534Z","l":"debug","m":"db","f":{"v":"[msg newVar i 0 plSQLArrays false typ 2010 natTyp 3000 sliceLen 1 bufSize 0 isSlice false]"}}
{"t":"2018-03-08T10:55:45.534Z","l":"debug","m":"db","f":{"v":"[msg set i 0 value int64=1386500022]"}}
{"t":"2018-03-08T10:55:45.535Z","l":"debug","m":"db","f":{"v":"[C dpiStmt_execute mode 32 colCount 0]"}}
{"t":"2018-03-08T11:00:45.535Z","l":"debug","m":"db","f":{"v":"[msg ExecContext context break returned]"}}
{"t":"2018-03-08T11:00:45.649Z","l":"debug","m":"db","f":{"v":"[msg st.Execute error ORA-01013: user requested cancel of current operation]"}}
{"t":"2018-03-08T11:00:45.649Z","l":"error","m":"Failed to update last scheduled date.","e":"dpiStmt_execute(mode=32 arrLen=-1): ORA-01013: user requested cancel of current operation"}
```
10:55 - db call started
11:00 - context timeout fires, break is called, break works
11:00 - ORA-01013: user requested cancel of current operation


While the above was the majority of errors.  I did receive a few of these:
```
{"t":"2018-03-08T02:16:02.980Z","l":"debug","m":"db.UpdateLastScheduleDate"}
{"t":"2018-03-08T02:16:02.980Z","l":"debug","m":"db","f":{"v":"[enter bindVars args [{ 1 1637100022}]]"}}
{"t":"2018-03-08T02:16:02.980Z","l":"debug","m":"db","f":{"v":"[msg bindVars i 0 in true out false value int64 1637100022]"}}
{"t":"2018-03-08T02:16:02.980Z","l":"debug","m":"db","f":{"v":"[doManyCount 1 arrLen -1 doExecMany false minArrLen maxArrLen]"}}
{"t":"2018-03-08T02:16:02.980Z","l":"debug","m":"db","f":{"v":"[msg newVar i 0 plSQLArrays false typ 2010 natTyp 3000 sliceLen 1 bufSize 0 isSlice false]"}}
{"t":"2018-03-08T02:16:02.980Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7f1e9c0ea600 typ 2010 natTyp 3000 sliceLen 1 bufSize 0 isArray 0 v <nil>]"}}
{"t":"2018-03-08T02:16:02.980Z","l":"debug","m":"db","f":{"v":"[msg set i 0 value int64=1637100022]"}}
{"t":"2018-03-08T02:16:02.980Z","l":"debug","m":"db","f":{"v":"[C dpiStmt_execute mode 32 colCount 0]"}}
{"t":"2018-03-08T02:17:15.033Z","l":"debug","m":"db","f":{"v":"[msg st.Execute error ORA-12609: TNS: Receive timeout occurred]"}}
{"t":"2018-03-08T02:17:15.033Z","l":"error","m":"Failed to update last scheduled date.","e":"dpiStmt_execute(mode=32 arrLen=-1): ORA-12609: TNS: Receive timeout occurred"}
```
and
```
{"t":"2018-03-08T09:57:43.075Z","l":"debug","m":"db.SelectMailBoxesToBeScheduled"}
{"t":"2018-03-08T09:57:43.075Z","l":"debug","m":"db","f":{"v":"[enter bindVars args []]"}}
{"t":"2018-03-08T09:57:43.075Z","l":"debug","m":"db","f":{"v":"[doManyCount 1 arrLen -1 doExecMany false minArrLen maxArrLen]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2010 natTyp 3000 sliceLen 256 bufSize 0 isArray 0 v <nil>]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2001 natTyp 3004 sliceLen 256 bufSize 1600 isArray 0 v <nil>]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2011 natTyp 3005 sliceLen 256 bufSize 0 isArray 0 v <nil>]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2011 natTyp 3005 sliceLen 256 bufSize 0 isArray 0 v <nil>]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2011 natTyp 3005 sliceLen 256 bufSize 0 isArray 0 v <nil>]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2001 natTyp 3004 sliceLen 256 bufSize 16 isArray 0 v <nil>]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2010 natTyp 3000 sliceLen 256 bufSize 0 isArray 0 v <nil>]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2010 natTyp 3004 sliceLen 256 bufSize 40 isArray 0 v <nil>]"}}
{"t":"2018-03-08T09:57:43.138Z","l":"debug","m":"db","f":{"v":"[C dpiConn_newVar conn 0x7fb9f00ea100 typ 2001 natTyp 3004 sliceLen 256 bufSize 512 isArray 0 v <nil>]"}}
{"t":"2018-03-08T10:04:29.194Z","l":"fatal","m":"Failed while running","e":"Next: ORA-12609: TNS: Receive timeout occurred"}
```
2:16 - db call started
2:17 - error while executing
2:17 - ORA-12609: TNS: Receive timeout occurred
(In other words, sometimes the receive timeout waited 6.5 minutes, and sometimes is waited only 1 minute.)



When I removed the `DISABLE_OOB` and `BREAK_POLL_SKIP` settings, my logs changed to show this instead:
```
{"t":"2018-03-09T07:54:17.309Z","l":"debug","m":"db.UpdateLastScheduleDate"}
{"t":"2018-03-09T07:54:17.309Z","l":"debug","m":"db","f":{"v":"[enter bindVars args [{ 1 1607000022}]]"}}
{"t":"2018-03-09T07:54:17.309Z","l":"debug","m":"db","f":{"v":"[msg bindVars i 0 in true out false value int64 1607000022]"}}
{"t":"2018-03-09T07:54:17.309Z","l":"debug","m":"db","f":{"v":"[doManyCount 1 arrLen -1 doExecMany false minArrLen maxArrLen]"}}
{"t":"2018-03-09T07:54:17.309Z","l":"debug","m":"db","f":{"v":"[msg newVar i 0 plSQLArrays false typ 2010 natTyp 3000 sliceLen 1 bufSize 0 isSlice false]"}}
{"t":"2018-03-09T07:54:17.309Z","l":"debug","m":"db","f":{"v":"[msg set i 0 value int64=1607000022]"}}
{"t":"2018-03-09T07:54:17.309Z","l":"debug","m":"db","f":{"v":"[C dpiStmt_execute mode 32 colCount 0]"}}
{"t":"2018-03-09T07:59:17.310Z","l":"debug","m":"db","f":{"v":"[msg ExecContext context break returned]"}}
{"t":"2018-03-09T08:14:06.304Z","l":"debug","m":"db","f":{"v":"[msg st.Execute error ORA-25408: can not safely replay call]"}}
{"t":"2018-03-09T08:14:06.304Z","l":"debug","m":"db","f":{"v":"[msg ExecContext closing done]"}}
{"t":"2018-03-09T08:14:06.304Z","l":"error","m":"Failed to update last scheduled date.","e":"dpiStmt_execute(mode=32 arrLen=-1): ORA-25408: can not safely replay call"}
```
7:54 - db call started
7:59 - context timeout fires, break is called, break does nothing
8:14 - ORA-25408: can not safely replay call
(I am not particularly sure why the ORA-25408 error happens around 10 minutes after the call is made, as this does not line up with any of my timeouts.)



Overall, both versions of sqlnet.ora gave me better results than not having it at all (where the db call hung forever).
However, it looks like the call to 'break' only works when `DISABLE_OOB = on`

The `BREAK_POLL_SKIP` setting is much trickier, because it is based on the number of packets sent, not on some sort of time interval.
Supposedly a low setting will increase CPU usage, but I honestly did not see any extra CPU usage due to setting this at 10.
My app is sending around 100 packets per minute, but that is the total, including traffic to SQS, an http liveness check, and a metrics endpoint.
If I had to guess, I'd say the actual packets sent to Oracle DB are around 5-30 per minute.

If I'd like to check for breaks every 5 minutes, that means a setting between 50-150 should be good.  I'm going to give this a try: `BREAK_POLL_SKIP = 100`

**Comment by cjbj** [comment]
@veqryn In https://github.com/oracle/odpi/issues/60#issuecomment-371237681 you give the Oracle versions.  What exact platforms are used by the DB and client?

**Comment by veqryn** [comment]
What do you mean?

**Comment by cjbj** [comment]
Linux (what distribution)? Windows?  32- or 64-bit?

**Comment by veqryn** [comment]
So the Oracle DB server is version 12.1.0.2 running on `3.8.13-118.16.4.el6uek.x86_64` (oracle linux 6 release 3, i think) on bare metal.  It is running in a local data center.

The clients are running in a Kubernetes 1.6.x cluster with CoreOS hosts, in a cloud provider's data center in a different state then the local data center (so there is probably a hypervisor too).
Our Golang Application is using Oracle Client 12.2.0.1 (from the docker container: container-registry.oracle.com/database/instantclient:12.2.0.1, which is Oracle Linux based).
Our Perl Application (which is seeing similar problems) is using Oracle Client 12.1.0.2, and is running on the perl 5.24.0 image in docker hub.

**Comment by cjbj** [comment]
Thank you.
